webapp/src/main/java/com/box/l10n/mojito/service/tm/TextUnitBatchOperator.java

```diff
@@ -0,0 +1,138 @@
+package com.box.l10n.mojito.service.tm;
+
+import com.box.l10n.mojito.service.tm.search.SearchType;
+import com.box.l10n.mojito.service.tm.search.StatusFilter;
+import com.box.l10n.mojito.service.tm.search.TextUnitDTO;
+import com.box.l10n.mojito.service.tm.search.TextUnitSearcher;
+import com.box.l10n.mojito.service.tm.search.TextUnitSearcherParameters;
+import com.google.common.base.Strings;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.util.ArrayList;
+import java.util.Iterator;
+import java.util.List;
+import java.util.function.BiConsumer;
+
+public class TextUnitBatchOperator {
+
+    static Logger logger = LoggerFactory.getLogger(TextUnitBatchOperator.class);
+
+    private final TextUnitSearcher searcher;
+    private final Long repositoryId;
+    private final List<String> localesTags;
+    private final String skipTextUnitsWithPattern;
+    private final String skipAssetsWithPathPattern;
+    private final int batchSize;
+
+    public TextUnitBatchOperator(TextUnitSearcher searcher,
+                                 Long repositoryId,
+                                 List<String> localesTags,
+                                 String skipTextUnitsWithPattern,
+                                 String skipAssetsWithPathPattern,
+                                 int batchSize) {
+        this.searcher = searcher;
+        this.repositoryId = repositoryId;
+        this.localesTags = localesTags;
+        this.skipTextUnitsWithPattern = skipTextUnitsWithPattern;
+        this.skipAssetsWithPathPattern = skipAssetsWithPathPattern;
+        this.batchSize = batchSize;
+    }
+
+    public void inBatches(BiConsumer<List<TextUnitDTO>, Integer> consumerSingulars,
+                          BiConsumer<List<TextUnitDTO>, Integer> consumerPlurals) {
+
+        int batch = 0;
+        TextUnitListIterator singularsIterator = new TextUnitListIterator(searcher, batchSize, repositoryId,
+                localesTags, skipTextUnitsWithPattern, skipAssetsWithPathPattern, true, true, null);
+
+        while (singularsIterator.hasNext()) {
+            consumerSingulars.accept(singularsIterator.next(), batch++);
+        }
+
+        TextUnitListIterator pluralsIterator = new TextUnitListIterator(searcher, batchSize, repositoryId,
+                localesTags, skipTextUnitsWithPattern, skipAssetsWithPathPattern, false, false, "%");
+
+        batch = 0;
+        while (pluralsIterator.hasNext()) {
+            consumerPlurals.accept(pluralsIterator.next(), batch++);
+        }
+    }
+
+    private static class TextUnitListIterator implements Iterator<List<TextUnitDTO>> {
```

Comment:
have you seen this class: `PageFetcherOffsetAndLimitSplitIterator`, it is used by the [Smartling impl](https://github.com/box/mojito/blob/0036896dd6710bf17fd38b03fa5f8fe98d6129b9/webapp/src/main/java/com/box/l10n/mojito/service/thirdparty/ThirdPartyTMSSmartling.java#L55) already, [drill down in the client](https://github.com/box/mojito/blob/41f4005831a77219fa0f519db3e7ae8f72017ece/webapp/src/main/java/com/box/l10n/mojito/smartling/SmartlingClient.java#L85) . it is useful for streaming data like that. 

I suspect having a stream will allow to remove `TextUnitBatchOperator`?

Reply:
I did took a quick peek at it, but considering we need a stateful iteration (we need to keep a reference to the `batchNumber` variable on both "consumers") and that we could lose the inherent streaming capabilities by forcing statefulness into a stateless stream, I moved to a classic Iterator to keep it simple.

Reply:
also thinking more about it, fetching the text unit DTOs by batch might not make a lot of sense here. Before it was done like that because of the http calls but that's gone now. So just batching when creating the file would simpler (and eventually long term this should rewritten in a more scalable way)

Reply:
Updated